lightning-persister/src/fs_store_async.rs

```diff
@@ -0,0 +1,155 @@
+		let key = key.to_string();
+		let this = Arc::clone(&self.inner);
+
+		Box::pin(async move {
+			tokio::task::spawn_blocking(move || {
+				this.read(&primary_namespace, &secondary_namespace, &key)
+			})
+			.await
+			.unwrap_or_else(|e| Err(lightning::io::Error::new(lightning::io::ErrorKind::Other, e)))
+		})
+	}
+
+	fn write(
+		&self, primary_namespace: &str, secondary_namespace: &str, key: &str, buf: &[u8],
+	) -> Pin<Box<dyn Future<Output = Result<(), lightning::io::Error>> + 'static + Send>> {
+		let primary_namespace = primary_namespace.to_string();
+		let secondary_namespace = secondary_namespace.to_string();
+		let key = key.to_string();
+		let buf = buf.to_vec();
+		let this = Arc::clone(&self.inner);
+
+		// Obtain a version number to retain the call sequence.
+		let version = self.version_counter.fetch_add(1, Ordering::SeqCst);
+
+		Box::pin(async move {
+			tokio::task::spawn_blocking(move || {
+				this.write_version(
+					&primary_namespace,
+					&secondary_namespace,
+					&key,
+					&buf,
+					Some(version),
+				)
+			})
+			.await
+			.unwrap_or_else(|e| Err(lightning::io::Error::new(lightning::io::ErrorKind::Other, e)))
+		})
+	}
+
+	fn remove(
+		&self, primary_namespace: &str, secondary_namespace: &str, key: &str, lazy: bool,
+	) -> Pin<Box<dyn Future<Output = Result<(), lightning::io::Error>> + 'static + Send>> {
+		let primary_namespace = primary_namespace.to_string();
+		let secondary_namespace = secondary_namespace.to_string();
+		let key = key.to_string();
+		let this = Arc::clone(&self.inner);
+
+		Box::pin(async move {
+			tokio::task::spawn_blocking(move || {
+				this.remove(&primary_namespace, &secondary_namespace, &key, lazy)
+			})
+			.await
+			.unwrap_or_else(|e| Err(lightning::io::Error::new(lightning::io::ErrorKind::Other, e)))
+		})
+	}
+
+	fn list(
+		&self, primary_namespace: &str, secondary_namespace: &str,
+	) -> Pin<Box<dyn Future<Output = Result<Vec<String>, lightning::io::Error>> + 'static + Send>> {
+		let primary_namespace = primary_namespace.to_string();
+		let secondary_namespace = secondary_namespace.to_string();
+		let this = Arc::clone(&self.inner);
+
+		Box::pin(async move {
+			tokio::task::spawn_blocking(move || this.list(&primary_namespace, &secondary_namespace))
+				.await
+				.unwrap_or_else(|e| {
+					Err(lightning::io::Error::new(lightning::io::ErrorKind::Other, e))
+				})
+		})
+	}
+}
+
+mod test {
+	#[tokio::test]
+	async fn read_write_remove_list_persist() {
+		use crate::{fs_store::FilesystemStore, fs_store_async::FilesystemStoreAsync};
+		use lightning::util::persist::KVStore;
+		use std::sync::Arc;
+
+		let mut temp_path = std::env::temp_dir();
+		temp_path.push("test_read_write_remove_list_persist");
+		let fs_store = Arc::new(FilesystemStore::new(temp_path));
+		let fs_store_async = FilesystemStoreAsync::new(Arc::clone(&fs_store));
+
+		let data1 = [42u8; 32];
+		let data2 = [43u8; 32];
+
+		let primary_namespace = "testspace_async";
```

Comment:
Tests execute in parallel. Needed to use a diff key here to avoid conflict with the sync test.